auth/tokencache/tokensource.go

```diff
@@ -0,0 +1,97 @@
+package tokencache
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	// refreshTimeout is the duration waited to receive a refresh token, before a new token is fetched.
+	// If the refresh takes longer than 5 seconds it is probably quicker to just fetch a new token (if possible).
+	refreshTimeout = 5 * time.Second
+
+	// minTTL is the minimum time that a cached token has to be valid for in order to be returned. 15 seconds should be
+	// sufficient for most calls that make use of a returned token.
+	minTTL = 15 * time.Second
+)
+
+// sourceType identities the type of token source.
+type sourceType = string
+
+// cachingTokenSource acts as a read-through cache for token information received from token sources and oauth configurations.
+type cachingTokenSource struct {
+	cacheFile        string
+	forceLogin       bool
+	sourceType       sourceType
+	sourceIdentifier string
+	oauthTokenSource oauth2.TokenSource
+	oauthConfig      oAuth2Config
+}
+
+// Token implements the oauth2.TokenSource interface. It will read cached tokens from a file and based on their validity
+// return, refresh or replace them.
+func (source *cachingTokenSource) Token() (*oauth2.Token, error) {
+	// Read the cache information from the file, if it exists
+	cachedTokens, err := readCache(source.cacheFile)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// Garbage collect expired tokens
+	cachedTokens.removeExpiredTokens()
+	_ = cachedTokens.write(source.cacheFile)
```

Comment:
I wasn't worried about this particular operation failing as it would just leave the expired tokens around and writing of new tokens failing should still be sufficient signal to know that writing is not possible.